src/xmerl_c14n.erl

```diff
@@ -0,0 +1,529 @@
+           end,
+    %% xmlns definitions
+    {Acc2, FinalActiveNS} =
+        if
+            not (Default =:= []) andalso not (Default =:= ParentDefault) ->
+                {["\"", xml_safe_string(Default, true), " xmlns=\"" | Acc1],
+                 [{default, Default} | NewActiveNS]};
+            not (Default =:= []) ->
+                {Acc1, [{default, Default} | NewActiveNS]};
+            true ->
+                {Acc1, NewActiveNS}
+        end,
+    Acc3 = lists:foldl(fun(Ns, AccIn) ->
+                               ["\"",xml_safe_string(
+                                       proplists:get_value(Ns, KnownNS, ""),
+                                       true),"=\"",Ns,":"," xmlns" | AccIn]
+                       end, Acc2, lists:sort(NewNS)),
+    %% any other attributes
+    Acc4 = lists:foldl(fun(Attr, AccIn) ->
+                               c14n(Attr, KnownNS, FinalActiveNS, Comments, [], AccIn)
+                       end, Acc3, Attrs),
+    %% close the opening tag
+    Acc5 = [">" | Acc4],
+
+    %% now accumulate all our children
+    Acc6 = lists:foldl(fun(Kid, AccIn) ->
+                               c14n(Kid, KnownNS, FinalActiveNS, Comments, [], AccIn)
+                       end, Acc5, Elem#xmlElement.content),
+
+    %% and finally add the close tag
+    case Elem#xmlElement.nsinfo of
+        {Ns, Name} ->
+            [">", Name, ":", Ns, "</" | Acc6];
+        _ ->
+            [">",atom_to_list(Elem#xmlElement.name),"</" | Acc6]
+    end;
+%% I do not give a shit
+c14n(_, _KnownNS, _ActiveNS, _Comments, _InclNs, Acc) ->
+    Acc.
+
+%% @doc Puts an XML document or element into canonical form, as a string.
+-spec c14n(XmlThing :: xml_thing()) -> string().
+c14n(Elem) ->
+    c14n(Elem, true).
+
+%% @doc Puts an XML document or element into canonical form, as a string.
+%%
+%% If the Comments argument is true, preserves comments in the output.
+-spec c14n(XmlThing :: xml_thing(), Comments :: boolean()) -> string().
+c14n(Elem, Comments) ->
+    c14n(Elem, Comments, []).
+
+%% @doc Puts an XML document or element into canonical form, as a string.
+%%
+%% If the Comments argument is true, preserves comments in the output. Any
+%% namespace prefixes listed in InclusiveNs will be left as they are and not
+%% modified during canonicalization.
+-spec c14n(XmlThing :: xml_thing(), Comments :: boolean(), InclusiveNs :: [string()]) -> string().
+c14n(Elem, Comments, InclusiveNs) ->
+    {NakedNss, _} = naked_nss(Elem, #{}, 0),
+    logger:debug("NakedNss: ~p~n", [lists:keysort(2, maps:to_list(NakedNss))]),
+    SortedNss = [ Prefix || {Prefix, _} <-
+                               lists:keysort(2, maps:to_list(NakedNss)) ],
+    %% FIXME: This still won't work. "ns14" < "ns4" in erlang and c14n/6
+    %% will put prefixes in the wrong order.
+    %% But we are saved by Skatteverket's removal of prefixes.
+    %% So we will leave this for a better day when we want to support
+    %% the complete standard.
```

Comment:
I mean this depends on who "we" are when using this library. Keep the FIXME about sorting not working but remove this reasoning saving one of our particular use cases.

Can we remove the sorting altogether here? Does it have any effect whatsoever given the caveats listed here?

Reply:
The standard says the namespace nodes are lexicographically sorted: https://www.w3.org/TR/2001/REC-xml-c14n-20010315#DocumentOrder so `ns14` before `ns4` sounds correct to me.

Reply:
That is the wrong standard.

Reply:
...but it probably applies since the excl standard doesn't say anything about ordering.

Reply:
And now that I reread the standard, only the attributes should be sorted with URI as primary key. Very confusing.